File: finalDS/queue.js
class Queue {
    constructor(head, end) {
        this.head = head;
        this.end = end;
    }
}
class node {
    constructor(data, next) {
        this.data = data;
        this.next = next;
    }
}

let myQueue = new Queue()

function gets(cb) {
    process.stdin.setEncoding('utf8');
    process.stdin.resume();
    process.stdin.on('data', function (chunk) {
        // console.log('start!');
        //去掉下一行可一直监听输入，即保持标准输入流为开启模式
        // process.stdin.pause();
        cb(chunk);
    });
    console.log('Please input string');
}

gets(function (m) {
    let result = m[0]
    if (m.length > 2) {
        console.log('U should input just one char.')
    } else if (result === '0') {
        // console.log(1)
        console.log('deQueue data is:', deQueue())
        // printQueue(myQueue.head, [])
    } else if (result === '@') {
        // console.log(2)
        printQueue(myQueue.head, [])
    } else if (result === '#') {
        console.log('The queue is ending:', printQueue(myQueue.head, []))
        process.stdin.emit('end');
    } else {
        // console.log(3)
        console.log('enqueue', result)
        enQueue(result)
    }
});

function deQueue() {
    if (myQueue.head === undefined) {
        console.log('There is an empty queue!!')
    } else {
        let d = myQueue.head.data
        myQueue.head = myQueue.head.next
        console.log(myQueue)
        return d
    }
}

function enQueue(text) {
    let thisNode = new node(text)

    if (myQueue.head === undefined) {
        myQueue.head = thisNode
        myQueue.end = thisNode
        myQueue.head.next = myQueue.end
    } else {
        myQueue.end.next = thisNode
        myQueue.end = thisNode
    }
    console.log(myQueue.head)
}

function printQueue(node, arr) {
    if (node === undefined) {
        console.log('This is an empty Queue!!')
        return
    }
    arr.push(node.data)
    if (node.next !== undefined) {
        return printQueue(node.next, arr)
    } else {
        console.log('Your Queue is:', arr.join(','))
    }
}

File: finalDS/queue.c
#include <stdio.h>
#include <stdlib.h>

char c;
typedef struct node {
  char data;
  struct node *next;
} Qnode, *Qlink;
typedef struct { Qnode *front, *rear; } linkqueue;

void Createqueue(linkqueue *q) {
  q->front = (Qlink)malloc(sizeof(Qnode));
  q->front->next = NULL;
  q->rear = q->front;
}
int Emptyqueue(linkqueue *q) {
  if (q->front == q->rear)
    return 1;
  else
    return 0;
}
void Enqueue(linkqueue *q, char e) {
  Qlink p;
  p = (Qlink)malloc(sizeof(Qnode));
  p->data = e;
  p->next = NULL;
  q->rear->next = p;
  q->rear = p;
}
char Dequeue(linkqueue *q) {
  Qlink p;
  if (Emptyqueue(q))
    return (NULL);
  else {
    p = q->front;
    q->front = p->next;
    free(p);
    return (q->front->data);
  }
}
int main() {
  linkqueue q;
  Createqueue(&q);
  while (c = getchar(), c - '#') {
    switch (c) {
      case '0': {
        if (Emptyqueue(&q)) {
          printf("Empty queue!\n");
          break;
        }
        printf("The char deQueued is : %c\n", Dequeue(&q));
        break;
      }
      case '@': {
        if (Emptyqueue(&q)) {
          printf("Empty queue!\n");
          break;
        }
        printf("The Queue is :");
        while (!Emptyqueue(&q)) {
          printf("%c", Dequeue(&q));
        }
        printf("\n");
        break;
      }
      case '\n': {
        break;
      }
      default: { Enqueue(&q, c); }
    }
  }
  return 0;
}